pkg/infra/iac2/templates_compiler.go

```diff
@@ -0,0 +1,189 @@
+package iac2
+
+import (
+	"embed"
+	_ "embed"
+	"fmt"
+	"github.com/klothoplatform/klotho/pkg/graph"
+	"github.com/klothoplatform/klotho/pkg/multierr"
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/pkg/errors"
+	"io"
+	"io/fs"
+	"reflect"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+type (
+	varNamer interface {
+		VariableName() string
+	}
+
+	resource struct {
+		hash    string
+		element any
+	}
+
+	templatesCompiler struct {
+		templates             fs.FS
+		resourceGraph         *graph.Directed[resource]
+		resources             map[any]resource
+		templatesByStructName map[string]ResourceCreationTemplate
+	}
+)
+
+var (
+	//go:embed templates/*/factory.ts
+	standardTemplates embed.FS
+
+	nonIdentifierChars = regexp.MustCompile(`\W`)
+)
+
+func CreateTemplatesCompiler() *templatesCompiler {
+	subTemplates, err := fs.Sub(standardTemplates, "templates")
+	if err != nil {
+		panic(err) // unexpected, since standardTemplates is statically built into klotho
+	}
+	return &templatesCompiler{
+		templates:             subTemplates,
+		resourceGraph:         graph.NewDirected[resource](),
+		resources:             make(map[any]resource),
+		templatesByStructName: make(map[string]ResourceCreationTemplate),
+	}
+}
+
+func (tc templatesCompiler) AddResource(v any) {
+	if _, exists := tc.resources[v]; exists {
+		return
+
+	}
+	res := resource{
+		hash:    fmt.Sprintf(`%x`, len(tc.resources)),
+		element: v,
+	}
+	tc.resources[v] = res
+	tc.resourceGraph.AddVertex(res)
+	for _, child := range getStructValues(v) {
+		if reflect.TypeOf(child).Kind() == reflect.Struct {
+			tc.AddResource(child)
+			childRes := tc.getResource(child)
+			tc.resourceGraph.AddEdge(res, childRes)
+		}
+	}
```

Comment:
why are the struct values dependencies? Maybe im a little confused on what `getStructValues` is returning

Reply:
If you have:

```
type Foo {
  hello string
  bar Bar
}

foo := Foo {
  hello: "world",
  bar: Bar{},
}
```

then `getStructValues(foo)` would return:

```
map[string]any {
  "hello": "world",
  "bar": Bar{},
}
```

This then means that `foo` depends on that `Bar{}`.

Reply:
by doing this secondary graph i think were doing extra work that we dont need to do. I see the usefulness in tracking the variable names, files, etc that the resource gets placed in, but if we recreate we could have a discrepancy between iac graph and provider resource graph

Reply:
Oh, I see what you mean. Yeah, this part can probably go away once I hook this up to the actual provider structs.

Basically, I needed a few changes to the `graph.Directed` to make the rendering work, but I didn't want to do that while your other PR was still up, in case we'd conflict (since I thought you were making some changes in that). So this is just a bit of scaffolding that will go away once we hook it up to the real provider structs.